stingray/deadtime/model.py

```diff
@@ -25,141 +99,192 @@ def r_det(td, r_i):
 @njit()
 def Gn(x, n):
     """Term in Eq. 34 in Zhang+95."""
-    s = 0
-    for l in range(0, n):
-        s += (n - l) / __FACTORIALS[l] * x**l
-    return np.exp(-x) * s
+    s = 0.0
 
+    for m in range(0, n):
+        new_val = e_m_x_x_over_factorial(x, m) * (n - m)
 
-@njit()
-def heaviside(x):
-    """Heaviside function. Returns 1 if x>0, and 0 otherwise.
-
-    Examples
-    --------
-    >>> heaviside(2)
-    1
-    >>> heaviside(-1)
-    0
-    """
-    if x >= 0:
-        return 1
-    else:
-        return 0
+        s += new_val
+        # The curve above has a maximum around x~l
+        if x != 0 and s > 0 and m > 2 * x and -np.log10(np.abs(new_val / s)) > PRECISION:
+            break
+
+    return s
 
 
 @njit()
 def h(k, n, td, tb, tau):
     """Term in Eq. 35 in Zhang+95."""
     # Typo in Zhang+95 corrected. k * tb, not k * td
-    if k * tb < n * td:
-        return 0
-    return k - n * (td + tau) / tb + tau / tb * Gn((k * tb - n * td) / tau, n)
+    factor = k * tb - n * td
+
+    if k * tb - n * td < 0:
+        return 0.0
+
+    val = k - n * (td + tau) / tb + tau / tb * Gn(factor / tau, n)
 
+    return val
 
-INFINITE = 100
+
+INFINITE = 699
 
 
 @njit()
 def A0(r0, td, tb, tau):
     """Term in Eq. 38 in Zhang+95."""
-    s = 0
-    for n in range(1, INFINITE):
+    s = 0.0
+    for n in range(1, int(max(2, tb / td * 2 + 1))):
         s += h(1, n, td, tb, tau)
 
     return r0 * tb * (1 + 2 * s)
 
 
 @njit()
-def A(k, r0, td, tb, tau):
+def A_single_k(k, r0, td, tb, tau):
     """Term in Eq. 39 in Zhang+95."""
     if k == 0:
         return A0(r0, td, tb, tau)
     # Equation 39
-    s = 0
-    for n in range(1, INFINITE):
-        s += h(k + 1, n, td, tb, tau) - 2 * h(k, n, td, tb, tau) + h(k - 1, n, td, tb, tau)
+    s = 0.0
+
+    for n in range(1, int(max(3, (k + 1) * tb / td * 2))):
+        new_val = h(k + 1, n, td, tb, tau) - 2 * h(k, n, td, tb, tau) + h(k - 1, n, td, tb, tau)
+
+        s += new_val
 
     return r0 * tb * s
 
 
-def check_A(rate, td, tb, max_k=100, save_to=None):
+def A(k, r0, td, tb, tau):
+    """Term in Eq. 39 in Zhang+95."""
+    if isinstance(k, Iterable):
+        return np.array([A_single_k(ki, r0, td, tb, tau) for ki in k])
+
+    return A_single_k(k, r0, td, tb, tau)
+
+
+def limit_A(rate, td, tb):
+    """Limit of A for k->infty, as per Eq. 43 in Zhang+95."""
+    r0 = r_det(td, rate)
+    return r0**2 * tb**2
+
+
+def check_A(rate, td, tb, max_k=100, save_to=None, linthresh=0.000001, rate_is_incident=True):
```

Comment:
I would write in the docstring that this function produces a plot. 
I would also copy and paste the description of the variables that are in common with the function pds_model_zhang in this docstring,  in order to make this function more independent.
I would also write at the end of the docstring "With this function is possible to determine how many inner loops k (limit_k in function pds_model_zhang) are necessary for a correct approximation of the dead time model" (or something on this line).
I think in this way the function check_A and pds_model_zhang are more connected (see also comment of the docstring of pds_model_zhang) 

I would apply the same comments on function check_B